File: apps/mobile/targets/positions-widget/Sources/Rendering/Medium/MediumPositionView.swift
import SwiftUI
import WidgetKit

struct MediumPositionView: View {
  let entry: PositionsEntry

  var body: some View {
    if let position = entry.position {
      content(position: position)
    } else {
      EmptyStateView(reason: entry.emptyReason ?? .notConfigured)
    }
  }

  @ViewBuilder
  private func content(position: WidgetPosition) -> some View {
    HStack(alignment: .top, spacing: Spacing.xl) {
      leftColumn(position: position)
        .frame(maxWidth: .infinity, alignment: .leading)
      Rectangle()
        .fill(Color.textPrimary.opacity(Opacity.strokeMuted))
        .frame(width: Sizing.Divider.thin)
      rightColumn(position: position)
        .layoutPriority(1)
    }
  }

  // MARK: - Left (identity + range)

  @ViewBuilder
  private func leftColumn(position: WidgetPosition) -> some View {
    VStack(alignment: .leading, spacing: Spacing.md) {
      pairRow(position: position)
      tagsRow(position: position)
      Spacer(minLength: 0)
      if case .uniswapV3(let payload) = position.widgetExtension, let range = payload.range {
        PriceRangeView(range: range)
      }
    }
  }

  @ViewBuilder
  private func pairRow(position: WidgetPosition) -> some View {
    HStack(alignment: .center, spacing: Spacing.xs) {
      Text("\(position.pair.sym0)/\(position.pair.sym1)")
        .widgetStyle(TypeScale.pairMd, scale: TextScale.aggressive)
        .accessibilityLabel("\(position.pair.sym0) \(position.pair.sym1) pair")
      Spacer(minLength: Spacing.md)
      ReverseButton(ref: position.ref, kind: .medium)
    }
  }

  @ViewBuilder
  private func tagsRow(position: WidgetPosition) -> some View {
    HStack(spacing: Spacing.xs) {
      ChainTag(chainID: position.chainId)
      if case .uniswapV3(let payload) = position.widgetExtension {
        MetaTag(text: payload.feeTierLabel)
      }
      StatusTag(status: position.status)
      Spacer(minLength: 0)
    }
    .accessibilityElement(children: .combine)
  }

  // MARK: - Right (value + fees)

  @ViewBuilder
  private func rightColumn(position: WidgetPosition) -> some View {
    let primary = position.primaryPrincipal
    let secondary = position.secondaryPrincipal
    VStack(alignment: .leading, spacing: Spacing.lg) {
      StatBlock(
        label: "Value",
        accent: .textPrimary,
        primary: TokenAmount(value: primary?.formatted, symbol: primary?.symbol),
        secondary: TokenAmount(value: secondary?.formatted, symbol: secondary?.symbol),
        density: .display
      )
      StatBlock(
        label: "Fees",
        accent: .brandPrimary,
        primary: TokenAmount(
          value: feeText(symbol: primary?.symbol, fees: position.fees),
          symbol: primary?.symbol
        ),
        secondary: TokenAmount(
          value: feeText(symbol: secondary?.symbol, fees: position.fees),
          symbol: secondary?.symbol
        ),
        density: .compact
      )
      Spacer(minLength: 0)
    }
  }

  private func feeText(symbol: String?, fees: [WidgetToken]) -> String? {
    guard let symbol else { return nil }
    return TokenStatHelper.feeString(for: symbol, in: fees)
  }
}

#if DEBUG
  #Preview("Medium — in range", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.inRange
  }

  #Preview("Medium — out of range", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.outOfRange
  }

  #Preview("Medium — edge left", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.edgeLeft
  }

  #Preview("Medium — empty", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.empty
  }

  #Preview("Medium — missing", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.missing
  }

  #Preview("Medium — tight stable (±50)", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.tightStable
  }

  #Preview("Medium — wide (±5000)", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.wideRange
  }

  #Preview("Medium — very wide (±20000)", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.veryWide
  }

  #Preview("Medium — far out above", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.farOutAbove
  }

  #Preview("Medium — far out below", as: .systemMedium) {
    PositionsWidget()
  } timeline: {
    PositionsEntry.farOutBelow
  }
#endif

File: apps/mobile/modules/widget-bridge/android/src/main/java/expo/modules/widgetbridge/widget/ui/MediumPositionView.kt
package expo.modules.widgetbridge.widget.ui

import androidx.compose.runtime.Composable
import androidx.compose.ui.unit.dp
import androidx.glance.GlanceModifier
import androidx.glance.action.Action
import androidx.glance.action.clickable
import androidx.glance.background
import androidx.glance.layout.Alignment
import androidx.glance.layout.Box
import androidx.glance.layout.Column
import androidx.glance.layout.ColumnScope
import androidx.glance.layout.Row
import androidx.glance.layout.Spacer
import androidx.glance.layout.fillMaxHeight
import androidx.glance.layout.fillMaxSize
import androidx.glance.layout.fillMaxWidth
import androidx.glance.layout.height
import androidx.glance.layout.padding
import androidx.glance.layout.width
import androidx.glance.unit.ColorProvider
import expo.modules.widgetbridge.widget.data.WidgetExtension
import expo.modules.widgetbridge.widget.data.WidgetPosition
import expo.modules.widgetbridge.widget.theme.Colors
import expo.modules.widgetbridge.widget.theme.Opacity
import expo.modules.widgetbridge.widget.theme.Sizing
import expo.modules.widgetbridge.widget.theme.Spacing
import expo.modules.widgetbridge.widget.theme.Typography
import expo.modules.widgetbridge.widget.ui.components.ChainTag
import expo.modules.widgetbridge.widget.ui.components.Density
import expo.modules.widgetbridge.widget.ui.components.MetaTag
import expo.modules.widgetbridge.widget.ui.components.PairRow
import expo.modules.widgetbridge.widget.ui.components.PriceRangeView
import expo.modules.widgetbridge.widget.ui.components.ReverseButtonKind
import expo.modules.widgetbridge.widget.ui.components.StatBlock
import expo.modules.widgetbridge.widget.ui.components.StatusTag
import expo.modules.widgetbridge.widget.ui.components.TokenAmount
import expo.modules.widgetbridge.widget.util.TokenStatHelper

private val RIGHT_COLUMN_WIDTH = 100.dp

@Composable
fun MediumPositionView(
  position: WidgetPosition?,
  reason: EmptyReason?,
  deepLinkAction: Action?,
) {
  if (position == null) {
    EmptyStateView(reason ?: EmptyReason.NotConfigured)
    return
  }
  Row(
    modifier = GlanceModifier.fillMaxSize().padding(Sizing.WidgetMetrics.contentMargin),
    verticalAlignment = Alignment.Vertical.CenterVertically,
  ) {
    Column(modifier = GlanceModifier.defaultWeight()) {
      LeftColumn(position, deepLinkAction)
    }
    Spacer(GlanceModifier.width(Spacing.lg))
    Box(
      modifier = GlanceModifier
        .width(Sizing.Divider.thin)
        .fillMaxHeight()
        .background(ColorProvider(Colors.textPrimary.copy(alpha = Opacity.strokeMuted))),
    ) {}
    Spacer(GlanceModifier.width(Spacing.lg))
    Column(modifier = GlanceModifier.width(RIGHT_COLUMN_WIDTH)) {
      RightColumn(position, deepLinkAction)
    }
  }
}

@Composable
private fun ColumnScope.LeftColumn(position: WidgetPosition, deepLinkAction: Action?) {
  // PairRow stays outside the deep-link clickable so its ReverseButton click
  // is not eaten by a wrapping handler.
  PairRow(
    sym0 = position.pair.sym0,
    sym1 = position.pair.sym1,
    positionRef = position.ref,
    pairStyle = Typography.pairMd,
    kind = ReverseButtonKind.Medium,
  )
  Spacer(GlanceModifier.height(Spacing.md))
  Box(
    modifier = GlanceModifier
      .defaultWeight()
      .fillMaxWidth()
      .let { m -> if (deepLinkAction != null) m.clickable(deepLinkAction) else m },
  ) {
    Column {
      TagsRow(position)
      val ext = position.extension as? WidgetExtension.UniswapV3
      if (ext?.range != null) {
        Spacer(GlanceModifier.height(Spacing.lg))
        PriceRangeView(range = ext.range)
      }
    }
  }
}

@Composable
private fun TagsRow(position: WidgetPosition) {
  val ext = position.extension as? WidgetExtension.UniswapV3
  Row(verticalAlignment = Alignment.Vertical.CenterVertically) {
    ChainTag(chainId = position.chainId)
    Spacer(GlanceModifier.width(Spacing.xs))
    if (ext != null) {
      MetaTag(text = ext.feeTierLabel)
      Spacer(GlanceModifier.width(Spacing.xs))
    }
    StatusTag(status = position.status)
    Spacer(GlanceModifier.defaultWeight())
  }
}

@Composable
private fun ColumnScope.RightColumn(position: WidgetPosition, deepLinkAction: Action?) {
  val primary = position.primaryPrincipal
  val secondary = position.secondaryPrincipal
  Box(
    modifier = GlanceModifier
      .fillMaxSize()
      .let { m -> if (deepLinkAction != null) m.clickable(deepLinkAction) else m },
  ) {
    Column {
      StatBlock(
        label = "Value",
        accent = Colors.textPrimary,
        primary = TokenAmount(value = primary?.formatted, symbol = primary?.symbol),
        secondary = TokenAmount(value = secondary?.formatted, symbol = secondary?.symbol),
        density = Density.Compact,
      )
      Spacer(GlanceModifier.height(Spacing.md))
      StatBlock(
        label = "Fees",
        accent = Colors.brandPrimary,
        primary = TokenAmount(
          value = TokenStatHelper.feeString(primary?.symbol, position.fees),
          symbol = primary?.symbol,
        ),
        secondary = TokenAmount(
          value = TokenStatHelper.feeString(secondary?.symbol, position.fees),
          symbol = secondary?.symbol,
        ),
        density = Density.Compact,
      )
    }
  }
}
